File: Battleships_api/Player.py
from .GameBoard import GameBoard
from .Ai import Ai
from . import References
import time
import random

class Player:
    def __init__(self, auto=False, test=False, aiLevel=0, boardSize=10):
        # TODO remove gameboards for computer players
        self.boardPrimary = GameBoard(boardSize)
        self.boardTracking = GameBoard(boardSize)
        # What remains of the fleet
        self.fleetSize = {\
            'A' : 5, \
            'B' : 4, \
            'C' : 3, \
            'S' : 3, \
            'D' : 2, \
            'shipsRemaining': 0 }
        # Location of the ships in the fleet
        self.fleetLocation = { 'Aircraft Carrier': [], \
            'Battleship': [], \
            'Cruiser': [], \
            'Submarine': [], \
            'Destroyer': [] }
        # TODO write a guard that stops shots being taken against a board that hasn't been setup
        #self.fleetLocationSet = False
        # records shots taken for checking if valid shot (memory inefficient, should use GameBoard)
        self.shotsTaken = []
        self.movesMade = 0
#        self.__setBoard(self.boardPrimary, auto=auto, test=test, randomise=randomise)
        self.autoPlayer = auto
        if self.autoPlayer:
            self.aIPlayer = Ai(aiLevel=aiLevel)
        self.setFleetLocation()
        self.latestShot = ()

    def getAutoPlayer(self):
        return self.autoPlayer

    def getLatestShot(self):
        if self.autoPlayer:
            return self.aIPlayer.getLatestShot()
        else:
            return self.latestShot

    def getBoard(self):
        return self.boardPrimary.getBoard()

    def getMovesMade(self):
        return self.movesMade

    def getTracking(self):
        return self.boardTracking.getBoard()

    def incoming(self, x, y):
        squareContents = self.boardPrimary.getSquare(x, y)
        # reverse lookup of symbol to shipname from dictionary
        shipName = next(key for key, value in References.symbols.items() if value == squareContents)
        # does not make clear what has been hit until ship has been destroyed
        if squareContents == ' ':
            self.boardPrimary.setSquare(x, y, References.getSymbols()['Miss'])
            return 'Miss', (x, y)
        elif squareContents != ' ' and self.fleetSize[squareContents] == 1:
             self.fleetSize[squareContents] -= 1
             self.fleetSize['shipsRemaining'] -= 1
             self.boardPrimary.setSquare(x, y, References.getSymbols()['Sunk'])
             self.__sinkShip(self.fleetLocation[shipName], self.boardPrimary)
             return shipName, self.fleetLocation[shipName] #currently needed for Ai to sink ships
        elif squareContents != ' ':
            self.fleetSize[squareContents] -= 1
            self.boardPrimary.setSquare(x, y, References.getSymbols()['Hit'])
            return 'Hit', (x, y)

    def takeShot(self, target, xCoord=False, yCoord=False):
        if self.autoPlayer:
            x, y = self.aIPlayer.takeShot()
            result = target.incoming(x, y)
            self.__recordShot(result, x, y)
        else:
            if (xCoord,yCoord) in self.shotsTaken:
                return -1
            self.latestShot = (xCoord, yCoord)
            self.shotsTaken.append((xCoord,yCoord))
            result = target.incoming(xCoord, yCoord)
            self.__recordShot(result, xCoord, yCoord)
        return result

    def __recordShot(self, result, x, y):
        if self.autoPlayer:
            self.aIPlayer.recordShot(result, x, y)
        elif result[0] in References.getShips():
            self.__sinkShip(result[1], self.boardTracking)
        else:
            self.boardTracking.setSquare(x, y, References.symbols[result[0]])

    def __sinkShip(self, locations, board):
        # iterate over length of ship replacing hit symbol with sunk symbol
        for i in locations:
            board.setSquare(i[0], i[1], References.getSymbols()['Sunk'])

    def __writeShip(self, grid, xCoord, yCoord, direction, shipName):
        # sends messages to board to set the locations of the fleet.
        for i in range(References.getShips()[shipName]):
            grid.setSquare(xCoord, yCoord, References.getSymbols()[shipName])
            #record ship coords.
            self.fleetLocation[shipName].append((xCoord, yCoord))
            if direction == 0:
                xCoord += 1
            elif direction == 1:
                yCoord += 1
        self.fleetSize['shipsRemaining'] += 1

    def __checkPlacement(self, grid, xCoord, yCoord, direction, shipName):
        if (xCoord+References.getShips()[shipName] > 10 and direction == 0)\
            or (yCoord+References.getShips()[shipName] > 10 and direction == 1):
            return False
        for i in range(References.getShips()[shipName]):
            if grid.getBoard()[yCoord][xCoord] != ' ':
                return False
            if direction == 0:
                xCoord += 1
            elif direction == 1:
                yCoord += 1
        return True

    def __placeShip(self, grid, xCoord, yCoord, direction, shipName):
        if self.__checkPlacement(grid, xCoord, yCoord, direction, shipName):
            self.__writeShip(grid, xCoord, yCoord, direction, shipName)
            return True
        return False

    def setFleetLocation(self, shipLocations=[], randomise=False):
        """ Takes a list containing the locations of the ship's coordinates in the order
        [[shipName, (xCoord, yCoord), direction],...]
        These need to have the correct number of locations in or it will not be accepted,
        valid ship placement will be also be checked and an error message returned if necessary.
        Assuming valid inputs, locations will be written to Player location dictionary, and
        a GameBoard instantiated. """
        # TODO employ defensive programming to protect from cheating
        if self.fleetSize['shipsRemaining'] == 5:
            return False
        elif self.autoPlayer or randomise:
            #print('setting random')
            self.__randomPlacement(self.boardPrimary)
            #print('auto player board setup done')
        elif len(shipLocations) >= 1:
            #print('setting from list')
            for eachShip in shipLocations:
                shipName, coords, direction = eachShip
                x,y = coords
                if not self.__placeShip(self.boardPrimary, x, y, direction, shipName):
                    #print('invalid placement in Player.setFleetLocation()')
                    return False
            return True
        pass

    def __randomPlacement(self, board):
        #print('randomPlacement')
        #if self.fleetLocationSet:
        #    return False
        #else:
        for eachShip in References.getShips().keys():
            placed = False
            while not placed:
                x = random.randrange(10)
                y = random.randrange(10)
                direction = random.randrange(2)
                placed = self.__placeShip(board, x, y, direction, eachShip)
        self.fleetLocationSet = True
        return True

File: battleships_java/battleships_java_rest/src/main/java/com/simalpas/battleships_java_rest/Player.java
package com.simalpas.battleships_java_rest;

import java.util.*;
import java.lang.Math;

/**
 * Object that represents a Player of the game Battleships.
 *
 * @author Simon Malpas
 */
public class Player
{

    private final GameBoard boardPrimary;
    private final GameBoard boardTracking;
    // the size of each ship remaining along with remaining ships
    private final Map<String, Integer> fleetSize;
    // the coordinates of the placed ships
    private final Map<String, List<Coord>> fleetLocation;
    // shotsTaken could not be used, instead the gameboard could be checked
    // to see if there has been a shot taken at that location. However whilst
    // this is slightly less memory efficient, it is simpler and a little faster
    private List<Coord> shotsTaken;
    private Coord latestShot;
    private int movesMade;
    private final boolean autoPlayer;
    private Ai aiBrain;

    /**
     * Constructor for a player
     *
     * @param auto is this player computer controlled
     * @param aiLevel 0=random, 1=random with shiptracking
     */
    protected Player(boolean auto, int aiLevel, int size)
    {
        // get out the boards
        boardPrimary = new GameBoard(size);
        boardTracking = new GameBoard(size);

        fleetSize = new HashMap<String, Integer>();

        this.fleetSize.put("A", 5);
        this.fleetSize.put("B", 4);
        this.fleetSize.put("C", 3);
        this.fleetSize.put("S", 3);
        this.fleetSize.put("D", 2);
        this.fleetSize.put("shipsRemaining", 5);

        fleetLocation = new HashMap<>();
        this.fleetLocation.put("Aircraft Carrier", new ArrayList<>());
        this.fleetLocation.put("Battleship", new ArrayList<>());
        this.fleetLocation.put("Cruiser", new ArrayList<>());
        this.fleetLocation.put("Submarine", new ArrayList<>());
        this.fleetLocation.put("Destroyer", new ArrayList<>());

        this.autoPlayer = auto;
        if (auto)
        {
            this.aiBrain = new Ai(aiLevel, this.boardPrimary);
//          Automatically setup any computer player boards
            this.setFleetLocation();
        }
    }

    /**
     * to allow testing
     *
     * @return
     */
    protected int getFleetSize(String shipSymbol)
    {
        return fleetSize.get(shipSymbol);
    }

    /**
     * to allow testing
     *
     * @return
     */
    protected Map<String, List<Coord>> getFleetLocation()
    {
        return fleetLocation;
    }

    protected char[][] getBoardPrimary()
    {
        return boardPrimary.getBoard();
    }

    protected char[][] getBoardTracking()
    {
        return boardTracking.getBoard();
    }

    protected int getMovesMade()
    {
        return movesMade;
    }

    protected boolean isAutoPlayer()
    {
        return autoPlayer;
    }

    protected Coord getLatestShot()
    {
        return this.latestShot;
    }

    /**
     *
     * @param x
     * @param y
     * @return single item map, key = result of shot/shipName if sunk value =
     * list of coordinates in the form of int arrays. single coord return is
     * shot just taken, multiple list entries if ship is sunk. Updates primary
     * board
     */
    protected Result incoming(Coord coord)
    {
        //conver char held on gameboard to String to match types in utility maps.
        String squareContents = String.valueOf(this.boardPrimary.getSquare(coord));

        // build map for return, input coords not used if ship is sunk
        // all this return faff is so that the AI player knows when it has
        // sunk a ship.
        Result result;

        // Shot misses
        if (squareContents.charAt(0) == ' ')
        {
            this.boardPrimary.setSquare(coord, References.getSymbol("Miss"));

            result = new Result("Miss", coord);
            return result;
        } // Shot sinks a ship
        else if (fleetSize.get(squareContents) == 1)
        {
            //reduce the size of the ship to 0
            fleetSize.put(squareContents, fleetSize.get(squareContents) - 1);
            // reduce the number of ships remaining by 1
            fleetSize.put("shipsRemaining", fleetSize.get("shipsRemaining") - 1);
            //find the name of the ship
            String shipName = References.getShipName(squareContents.charAt(0));

            // return the name of the ship and the locations of each part of
            // of the ship, first entry in list of coords in Result object is
            // incoming shot
            List<Coord> coords = new ArrayList<>();
            coords.add(coord);
            for (Coord eachLocation : fleetLocation.get(squareContents))
            {
                coords.add(eachLocation);
            }
            result = new Result(shipName, coords);
            // change the symbols on the board to sunk
            this.sinkShip(result, this.boardPrimary);

            return result;
        } else

        {
            result = new Result("Hit", coord);
        }
        return result;
    }

    private void sinkShip(Result inputResult, GameBoard gb)
    {
        //loop through fleetLocation to sink ship
        for (Coord eachCoord : inputResult.getAllCoords())
        {
            gb.setSquare(eachCoord, References.getSymbol("Sunk"));
        }
    }

    /**
     * Overloaded method No input coords required if ai player.
     * records the aiBrain's shot attempt
     *
     * @param target Player object to shoot at
     * @return
     */
    protected Result takeShot(Player target)
    {
        Result result;
        if (autoPlayer)
        {
            // aiBrain will not return a duplicate coord, no need to check
            Coord coords = aiBrain.takeShot();
            this.latestShot = coords;
            result = target.incoming(coords);
            this.recordShotTaken(result);
            return result;
        }
        result = new Result();
        return result;
    }

    /**
     * Overloaded method Covers the case where coordinates are provided, either
     * human player or external ai brain. checks it is a valid shot. finds out
     * the result of the shot. records coords of shot taken. updates tracking
     * board (sinking if necessary). increases moves made.
     *
     *
     * @param result
     * @param x x coordinate for shot
     * @param y y coordinate for shot
     * @return single entry map with shipName/outcome, list of int arrays with
     * either locations of sunken ship or shot just taken
     */
    protected Result takeShot(Player target, Coord coords)
    {
        Result result;
        // contains method uses overridden equals method to compare coords
        if (this.shotsTaken.contains(coords))
        {
            result = new Result("invalid", coords);
            return result;
        }
        //find results of shot
        // 1st coord in result is latest shot
        result = target.incoming(coords);
        String resultName = result.getResultName();
        // update tracking board and metrics
        this.recordShotTaken(result);
        return result;
    }

    /**
     * Utility method. lets the aiBrain know whats going on so it can track
     * unsunk ships Updates tracking board, updates latest shot,
     *
     * @param inputResult information on what happened.
     */
    private void recordShotTaken(Result inputResult)
    {
        if (this.autoPlayer)
        {
            this.aiBrain.recordShot(inputResult);
        }
        String resultName = inputResult.getResultName();
        Coord shot = inputResult.get1stCoord();
        // check if a ship was sunk
        if (References.SHIPS.containsKey(resultName))
        {
            //if it was update the tracking board
            this.sinkShip(inputResult, this.boardTracking);
        } // if not: hit or miss
        else
        {
            this.boardTracking.setSquare(shot, References.getSymbol(resultName));
        }
        this.latestShot = shot;
        this.shotsTaken.add(shot);
    }

    /**
     * Sets the board according to ship location, and records the coords.
     *
     * @param coords origin of the ship
     * @param direction 0=horizontal 1=vertical
     * @param shipName which ship to place
     */
    private void writeShip(Coord coords, int direction, String shipName)
    {
        // unpack coords
        int x = coords.getX();
        int y = coords.getY();
        // iterate for the length of the ship
        for (int i = References.SHIPS.get(shipName); i > 0; i--)
        {
            // set the square
            Coord eachCoord = new Coord(x, y);
            this.boardPrimary.setSquare(eachCoord, References.getSymbol(shipName));
            // record the coords in fleetLocation
            this.fleetLocation.get(shipName).add(eachCoord);
            // increment the appropriate coordinate according to direction
            if (direction == 0)
            {
                x++;
            } else if (direction == 1)
            {
                y++;
            }
        }
        // exit loop and increment ships remaining
        this.fleetSize.put("shipsRemaining", this.fleetSize.get("shipsRemaining") + 1);
    }

    /**
     * Ensures coordinates are in range, and there is no overlap.
     *
     * @param coords origin point of the ship
     * @param direction 0/1 the direction the ship will be placed horz/vert
     * @param shipName the name of the ship being checked
     * @return if placement is valid
     */
    private boolean checkPlacement(Coord coords, int direction, String shipName)
    {
        int x = coords.getX();
        int y = coords.getY();
        if ((x + References.SHIPS.get(shipName) > 10)
                || (y + References.SHIPS.get(shipName) > 10))
        {
            return false;
        }
        Coord tempCoord;
        for (int i = References.SHIPS.get(shipName); i > 0; i--)
        {
            tempCoord = new Coord(x, y);
            if (this.boardPrimary.getSquare(tempCoord) != ' ')
            {
                return false;
            }
            if (direction == 0)
            {
                x++;
            }
            if (direction == 1)
            {
                y++;
            }
        }
        return true;
    }

    /**
     * Check proposed placement, and writes the location if valid
     *
     * @param coords origin of the ship
     * @param direction 0=horizontal 1=vertical
     * @param shipName name of the ship being placed
     * @return successful or not
     */
    private boolean placeShip(Coord coords, int direction, String shipName)
    {
        if (this.checkPlacement(coords, direction, shipName))
        {
            this.writeShip(coords, direction, shipName);
            return true;
        }
        return false;
    }

    /**
     * Zero arg call places ships at random in valid positions
     *
     * @return successful or not
     */
    protected boolean setFleetLocation()
    {
        return this.randomPlacement();
    }

    /**
     * Sets the fleet if valid
     *
     * @param proposedFleet Map made up of ShipName, list of Integers in the
     * following order - OriginXCoord, OriginYCoord, direction (0/1). Where
     * 0=horiz, 1=vert.
     */
    protected boolean setFleetLocation(Map<String, Integer[]> proposedFleet)
    {
        // iterate through Map proposedFleet
        for (Map.Entry<String, Integer[]> eachEntry : proposedFleet.entrySet())
        {
            // check ship hasn't been placed already
            if (this.fleetLocation.get(eachEntry.getKey()).size() > 0)
            {
                return false;
            }
            // unpack values to pass to placeShip
            int tempX = eachEntry.getValue()[0];
            int tempY = eachEntry.getValue()[1];
            int tempDirection = eachEntry.getValue()[2];
            String tempShipName = eachEntry.getKey();
            Coord tempCoord = new Coord(tempX, tempY);
            // try and place the ship, returning false if not possible
            return this.placeShip(tempCoord, tempDirection, tempShipName);
        }
        return false;
    }

    /**
     * Places ships at random for a fast setup
     *
     * @return successful or not
     */
    private boolean randomPlacement()
    {
        // iterate through all ships in References.SHIPS
        for (Map.Entry<String, Integer> eachShip : References.SHIPS.entrySet())
        {
            // check the ship hasn't already been placed
            if (this.fleetLocation.get(eachShip.getKey()).size() > 0)
            {
                // do nothing
            } else
            {
                boolean placed = false;
                // loop whilst the ship isn't placed
                while (!placed)
                {
                    // pick random coordinates and direction
                    // casting to int rounds down
                    int x = (int) (Math.random() * this.boardPrimary.getSIZE());
                    int y = (int) (Math.random() * this.boardPrimary.getSIZE());
                    int direction = (int) (Math.random() * 2);
                    Coord tempCoord = new Coord(x,y);
                    // check if it is a valid origin, and place it if so.
                    // update loop var to exit if successful
                    placed = this.placeShip(tempCoord, direction, eachShip.getKey());
//                    {
//                        // increase the number of shipsRemaining
//                        this.fleetSize.put("shipsRemaining", this.fleetSize.get("shipsRemaining")+1);
//                        // update placed to exit loop
//                        placed =true;
//                    }
                }
            }
        }
        // return whether the shipsRemaining == 5
        return (this.fleetSize.get("shipsRemaining") == 5);
    }
}
